tensorflow/contrib/opt/python/training/irprop_plus.py

```diff
@@ -0,0 +1,319 @@
+  algorithm is empirically found to be faster and more robust than the
+  [standard variant](RpropMinusOptimizer.md).
+  See [Resilient Backpropagation (Rprop) for Batch-learning in
+  TensorFlow](https://openreview.net/forum?id=r1R0o7yDz)
+  for details and references.
+  """
+
+  # Values for gate_gradients, taken from base class.
+  GATE_NONE = 0
+  GATE_OP = 1
+  GATE_GRAPH = 2
+
+  def __init__(self,
+               eta_minus=0.5,
+               eta_plus=1.2,
+               delta_zero=0.5,
+               delta_min=1e-6,
+               delta_max=50,
+               use_locking=False, name="IRpropPlusOptimizer"):
+    """Constructs a new IRpropPlusOptimizer object.
+
+    The pseudocode of the algorithm can be found in the articles
+    [Empirical evaluation of the improved Rprop learning algorithms](https://doi.org/10.1016/S0925-2312(01)00700-7)
+    and [Resilient Backpropagation (Rprop) for Batch-learning in TensorFlow](https://openreview.net/forum?id=r1R0o7yDz).
+
+    Initialization:
+
+    ```
+    old_grad <- 0 (Initialize the gradient from the previous iteration g{t-1})
+    delta_update <- delta_zero (Initialize step-size)
+    t <- 0 (Initialize iteration counter)
+    ```
+
+    The following update rule is performed for each individual objective
+    variable (e.g., weight), where `g{t}` denotes the partial derivative of the
+    objective function with respect to the objective variable at iteration `t`:
+
+    ```
+    t <- t + 1
+    grad_sign <- sign(g{t} * g{t-1})
+    if (grad_sign > 0)
+      delta_update{t} <- min(eta_plus * delta_update{t-1}, delta_max)
+      variable{t+1} <- variable{t} -sign(g{t}) * delta_update{t}
+    else if (grad_sign < 0)
+      delta_update{t} <- max(eta_minus * delta_update{t-1}, delta_min)
+      if (error{t} > error{t-1})
+        variable{t+1} <- variable{t-1}
+      g{t} = 0
+    else
+      variable{t+1} <- variable{t} - sign(g{t}) * delta_update{t}
+
+    ```
+
+    Args:
+      eta_minus: Step-size decrease factor.
+      eta_plus: Step-size increase factor.
+      delta_zero: Initial step-size.
+      delta_min: Lower bound on step-size.
+      delta_max: Upper bound on step-size.
+      use_locking: If True, use locks for update operations.
+      name: Optional name for the operations created when applying gradients.
+         Defaults to "IRpropPlusOptimizer".
+   """
+
+    super(IRpropPlusOptimizer, self).__init__(use_locking, name)
+
+    # Init parameters
+    self._eta_minus = eta_minus
+    self._eta_plus = eta_plus
+    self._delta_zero = delta_zero
+    self._delta_min = delta_min
+    self._delta_max = delta_max
+
+    # Tensor versions of the constructor arguments, created in _prepare().
+    self._eta_minus_t = None
+    self._eta_plus_t = None
+    # self._delta_zero_t = None
+    self._delta_min_t = None
+    self._delta_max_t = None
+
+    # Error tensors
+    self._error = variable_scope.variable(0.0,
```

Comment:
Defer creation of tensors or variables to prepare, since optimizers are supposed to be reusable across multiple graphs.

Also use _create_non_slot_variable.

I'd prefer it if you inherited from OptimizerV2 instead of Optimizer as we'll do the switch soon.

Reply:
I've inherited from OptimizerV2 and create the error variables using  _create_non_slot_variable. However we still need a variable that can hold the loss value at the current iteration before calling apply_gradients, in order to feed the error in the update step. That is also because _create_vars is called in apply_gradients method.